File: src/v3/runtime/limits.ts
import { DiceRollError } from '../errors.js';

export interface DiceLimits {
  readonly maxInputLength: number;
  readonly maxAstDepth: number;
  readonly maxAstNodes: number;
  readonly maxRolls: number;
  readonly maxInitialDice: number;
  readonly maxGeneratedDice: number;
  readonly maxRandomCalls: number;
  readonly maxEvents: number;
  readonly maxSides: number;
  readonly maxSeedLength: number;
  readonly maxModifierSteps: number;
  readonly maxResolvedGroups: number;
  readonly maxResultItems: number;
  readonly maxOutputLength: number;
}

export type DiceLimitOverrides = Partial<DiceLimits>;

export const DEFAULT_DICE_LIMITS: DiceLimits = Object.freeze({
  maxInputLength: 4_096,
  maxAstDepth: 64,
  maxAstNodes: 10_000,
  maxRolls: 100,
  maxInitialDice: 10_000,
  maxGeneratedDice: 20_000,
  maxRandomCalls: 100_000,
  maxEvents: 100_000,
  maxSides: 0x1_0000_0000,
  maxSeedLength: 1_024,
  maxModifierSteps: 100_000,
  maxResolvedGroups: 100_000,
  maxResultItems: 250_000,
  maxOutputLength: 1_000_000,
});

type DiceLimitName = keyof DiceLimits;

function validateLimit(name: DiceLimitName, value: number): number {
  if (!Number.isSafeInteger(value) || value < 1) {
    throw new DiceRollError(`${name} must be a positive safe integer`, {
      code: 'INVALID_LIMIT',
      details: {
        limit: name,
        value: Number.isFinite(value) ? value : String(value),
      },
    });
  }

  return value;
}

function readOverride(
  overrides: DiceLimitOverrides,
  name: DiceLimitName,
  fallback: number,
): number {
  let value: unknown;
  try {
    value = overrides[name];
  } catch {
    throw new DiceRollError(`Unable to read ${name}`, {
      code: 'INVALID_LIMIT',
      details: { limit: name },
    });
  }
  if (value === undefined) {
    return fallback;
  }
  if (typeof value !== 'number') {
    throw new DiceRollError(`${name} must be a number`, {
      code: 'INVALID_LIMIT',
      details: { limit: name, valueType: typeof value },
    });
  }
  return validateLimit(name, value);
}

function assertOverridesObject(overrides: unknown): asserts overrides is DiceLimitOverrides {
  try {
    if (typeof overrides !== 'object' || overrides === null || Array.isArray(overrides)) {
      throw new DiceRollError('Dice limit overrides must be a non-null object', {
        code: 'INVALID_LIMIT',
      });
    }
  } catch (error: unknown) {
    if (error instanceof DiceRollError) {
      throw error;
    }
    throw new DiceRollError('Dice limit overrides could not be read', {
      code: 'INVALID_LIMIT',
    });
  }
}

function resolveOverride(
  engineLimits: DiceLimits,
  overrides: DiceLimitOverrides,
  name: DiceLimitName,
): number {
  const value = readOverride(overrides, name, engineLimits[name]);
  if (value > engineLimits[name]) {
    throw new DiceRollError(`${name} cannot exceed the engine cap`, {
      code: 'INVALID_LIMIT',
      details: {
        limit: name,
        requested: value,
        engineCap: engineLimits[name],
      },
    });
  }
  return value;
}

function freezeLimits(limits: DiceLimits): DiceLimits {
  return Object.freeze(limits);
}

/** Creates the immutable caps owned by a DiceEngine instance. */
export function createDiceLimits(overrides: unknown = {}): DiceLimits {
  assertOverridesObject(overrides);
  return freezeLimits({
    maxInputLength: readOverride(overrides, 'maxInputLength', DEFAULT_DICE_LIMITS.maxInputLength),
    maxAstDepth: readOverride(overrides, 'maxAstDepth', DEFAULT_DICE_LIMITS.maxAstDepth),
    maxAstNodes: readOverride(overrides, 'maxAstNodes', DEFAULT_DICE_LIMITS.maxAstNodes),
    maxRolls: readOverride(overrides, 'maxRolls', DEFAULT_DICE_LIMITS.maxRolls),
    maxInitialDice: readOverride(overrides, 'maxInitialDice', DEFAULT_DICE_LIMITS.maxInitialDice),
    maxGeneratedDice: readOverride(
      overrides,
      'maxGeneratedDice',
      DEFAULT_DICE_LIMITS.maxGeneratedDice,
    ),
    maxRandomCalls: readOverride(overrides, 'maxRandomCalls', DEFAULT_DICE_LIMITS.maxRandomCalls),
    maxEvents: readOverride(overrides, 'maxEvents', DEFAULT_DICE_LIMITS.maxEvents),
    maxSides: readOverride(overrides, 'maxSides', DEFAULT_DICE_LIMITS.maxSides),
    maxSeedLength: readOverride(overrides, 'maxSeedLength', DEFAULT_DICE_LIMITS.maxSeedLength),
    maxModifierSteps: readOverride(
      overrides,
      'maxModifierSteps',
      DEFAULT_DICE_LIMITS.maxModifierSteps,
    ),
    maxResolvedGroups: readOverride(
      overrides,
      'maxResolvedGroups',
      DEFAULT_DICE_LIMITS.maxResolvedGroups,
    ),
    maxResultItems: readOverride(
      overrides,
      'maxResultItems',
      DEFAULT_DICE_LIMITS.maxResultItems,
    ),
    maxOutputLength: readOverride(
      overrides,
      'maxOutputLength',
      DEFAULT_DICE_LIMITS.maxOutputLength,
    ),
  });
}

/**
 * Resolves per-call limits against immutable engine caps.
 * A call may lower a cap, but attempting to raise one is a configuration error.
 */
export function resolveDiceLimits(
  engineLimits: DiceLimits,
  callOverrides: unknown = {},
): DiceLimits {
  assertOverridesObject(callOverrides);

  return freezeLimits({
    maxInputLength: resolveOverride(engineLimits, callOverrides, 'maxInputLength'),
    maxAstDepth: resolveOverride(engineLimits, callOverrides, 'maxAstDepth'),
    maxAstNodes: resolveOverride(engineLimits, callOverrides, 'maxAstNodes'),
    maxRolls: resolveOverride(engineLimits, callOverrides, 'maxRolls'),
    maxInitialDice: resolveOverride(engineLimits, callOverrides, 'maxInitialDice'),
    maxGeneratedDice: resolveOverride(engineLimits, callOverrides, 'maxGeneratedDice'),
    maxRandomCalls: resolveOverride(engineLimits, callOverrides, 'maxRandomCalls'),
    maxEvents: resolveOverride(engineLimits, callOverrides, 'maxEvents'),
    maxSides: resolveOverride(engineLimits, callOverrides, 'maxSides'),
    maxSeedLength: resolveOverride(engineLimits, callOverrides, 'maxSeedLength'),
    maxModifierSteps: resolveOverride(engineLimits, callOverrides, 'maxModifierSteps'),
    maxResolvedGroups: resolveOverride(engineLimits, callOverrides, 'maxResolvedGroups'),
    maxResultItems: resolveOverride(engineLimits, callOverrides, 'maxResultItems'),
    maxOutputLength: resolveOverride(engineLimits, callOverrides, 'maxOutputLength'),
  });
}

File: go/limits.go
package dicecore

const maxSafeInteger int64 = 9_007_199_254_740_991

// DiceLimits are value-owned caps. A caller cannot mutate an existing budget by
// changing the struct from which it was created.
type DiceLimits struct {
	MaxInputLength    int64 `json:"maxInputLength"`
	MaxAstDepth       int64 `json:"maxAstDepth"`
	MaxAstNodes       int64 `json:"maxAstNodes"`
	MaxRolls          int64 `json:"maxRolls"`
	MaxInitialDice    int64 `json:"maxInitialDice"`
	MaxGeneratedDice  int64 `json:"maxGeneratedDice"`
	MaxRandomCalls    int64 `json:"maxRandomCalls"`
	MaxEvents         int64 `json:"maxEvents"`
	MaxSides          int64 `json:"maxSides"`
	MaxSeedLength     int64 `json:"maxSeedLength"`
	MaxModifierSteps  int64 `json:"maxModifierSteps"`
	MaxResolvedGroups int64 `json:"maxResolvedGroups"`
	MaxResultItems    int64 `json:"maxResultItems"`
	MaxOutputLength   int64 `json:"maxOutputLength"`
}

// DiceLimitOverrides distinguishes an omitted field from an invalid zero cap.
type DiceLimitOverrides map[string]int64

func DefaultDiceLimits() DiceLimits {
	return DiceLimits{4096, 64, 10_000, 100, 10_000, 20_000, 100_000, 100_000,
		1 << 32, 1024, 100_000, 100_000, 250_000, 1_000_000}
}

func TrustedServerDiceLimits() DiceLimits {
	return DiceLimits{2048, 64, 5000, 100, 5000, 10_000, 50_000, 50_000,
		1 << 32, 1024, 50_000, 50_000, 125_000, 500_000}
}

func UntrustedServerDiceLimits() DiceLimits {
	return DiceLimits{1000, 32, 1000, 50, 500, 1000, 5000, 10_000,
		1_000_000, 256, 5000, 10_000, 25_000, 100_000}
}

type limitField struct {
	name  string
	value *int64
}

func limitFields(limits *DiceLimits) []limitField {
	return []limitField{
		{"maxInputLength", &limits.MaxInputLength}, {"maxAstDepth", &limits.MaxAstDepth},
		{"maxAstNodes", &limits.MaxAstNodes}, {"maxRolls", &limits.MaxRolls},
		{"maxInitialDice", &limits.MaxInitialDice}, {"maxGeneratedDice", &limits.MaxGeneratedDice},
		{"maxRandomCalls", &limits.MaxRandomCalls}, {"maxEvents", &limits.MaxEvents},
		{"maxSides", &limits.MaxSides}, {"maxSeedLength", &limits.MaxSeedLength},
		{"maxModifierSteps", &limits.MaxModifierSteps}, {"maxResolvedGroups", &limits.MaxResolvedGroups},
		{"maxResultItems", &limits.MaxResultItems}, {"maxOutputLength", &limits.MaxOutputLength},
	}
}

func validateLimit(name string, value int64) error {
	if value < 1 || value > maxSafeInteger {
		return newDiceError("INVALID_LIMIT", name+" must be a positive safe integer", "",
			map[string]any{"limit": name, "value": value})
	}
	return nil
}

// Validate checks caps assembled directly as a Go struct.
func (limits DiceLimits) Validate() error {
	for _, field := range limitFields(&limits) {
		if err := validateLimit(field.name, *field.value); err != nil {
			return err
		}
	}
	return nil
}

// CreateDiceLimits creates engine caps from defaults. Unlike per-call overrides,
// engine caps may be raised to another positive, JavaScript-safe integer.
func CreateDiceLimits(overrides DiceLimitOverrides) (DiceLimits, error) {
	limits := DefaultDiceLimits()
	for _, field := range limitFields(&limits) {
		if value, ok := overrides[field.name]; ok {
			if err := validateLimit(field.name, value); err != nil {
				return DiceLimits{}, err
			}
			*field.value = value
		}
	}
	return limits, nil
}

// ResolveDiceLimits allows a call to lower engine caps, never to raise them.
func ResolveDiceLimits(engine DiceLimits, overrides DiceLimitOverrides) (DiceLimits, error) {
	if err := engine.Validate(); err != nil {
		return DiceLimits{}, err
	}
	for _, field := range limitFields(&engine) {
		if value, ok := overrides[field.name]; ok {
			if err := validateLimit(field.name, value); err != nil {
				return DiceLimits{}, err
			}
			if value > *field.value {
				return DiceLimits{}, newDiceError("INVALID_LIMIT", field.name+" cannot exceed the engine cap", "",
					map[string]any{"limit": field.name, "requested": value, "engineCap": *field.value})
			}
			*field.value = value
		}
	}
	return engine, nil
}
